af_ktls.c

```diff
@@ -883,15 +885,17 @@ static int tls_sendmsg(struct socket *sock, struct msghdr *msg, size_t size)
 
 	tls_make_aad(tsk, 0, tsk->aad_send, size, tsk->iv_send);
 
-	ret = af_alg_make_sg(&tsk->sgl_send, &msg->msg_iter, size);
-	if (ret < 0)
-		goto send_end;
-
-	sg_unmark_end(&tsk->sgaad_send[1]);
-	sg_chain(tsk->sgaad_send, 2, tsk->sgl_send.sg);
-
-	sg_unmark_end(tsk->sgl_send.sg + tsk->sgl_send.npages - 1);
-	sg_chain(tsk->sgl_send.sg, tsk->sgl_send.npages + 1, tsk->sgtag_send);
+    while (iov_iter_count(&msg->msg_iter)) {
+        size_t seglen = iov_iter_count(&msg->msg_iter);
+        int len = af_alg_make_sg(&tsk->sgl_send[cnt], &msg->msg_iter, seglen);
+        if (len < 0)
+            goto send_end;
+        ret += len;
+        if (cnt)
+            af_alg_link_sg(&tsk->sgl_send[cnt-1], &tsk->sgl_send[cnt]);
+        iov_iter_advance(&msg->msg_iter, len);
+        cnt++;
+    }
```

Comment:
This could be wrapped to a separate function. ```af_alg_make_sg``` will be removed from AF_ALG (see https://github.com/fridex/af_ktls/issues/27), so we can introduce a helper that would cover iterating over iovec and use refactored ```af_alg_make_sg```. Nevertheless this approach is not not needed in my opinion - see comment in this pull request.